File: 0019-remove-nth-node-from-end-of-list/0019-remove-nth-node-from-end-of-list.ts
/**
 * Definition for singly-linked list.
 * class ListNode {
 *     val: number
 *     next: ListNode | null
 *     constructor(val?: number, next?: ListNode | null) {
 *         this.val = (val===undefined ? 0 : val)
 *         this.next = (next===undefined ? null : next)
 *     }
 * }
 */

function len(head: ListNode | null): number {
	let sz = 0
	while (!!head) {
		sz += 1
		head = head.next
	}
	return sz
}

function remove(head: ListNode | null, nth: number): ListNode | null {
	const dummy = new ListNode(-1, head)
	let [prev, it]: [ListNode | null, ListNode | null] = [dummy, head]

	for (let _ = 0; _ < nth; _++) {
		[prev, it] = [it, it!!.next]
	}
	prev!!.next = it!!.next
	return dummy.next
}

function removeNthFromEnd(head: ListNode | null, n: number): ListNode | null {
	const sz = len(head)
	const nth = sz - n

	return remove(head, nth)
};

File: 0019-remove-nth-node-from-end-of-list/0019-remove-nth-node-from-end-of-list.py
# Definition for singly-linked list.
# class ListNode:
#     def __init__(self, val=0, next=None):
#         self.val = val
#         self.next = next
class Solution:
    def len(self, root):
        sz = 0
        while root:
            sz += 1
            root = root.next
        return sz
    
    def remove(self, head, n):
        dummy = ListNode(val=-1, next=head)
        prev, it = dummy, head

        for _ in range(n):
            prev, it = it, it.next
        prev.next = it.next
        return dummy.next

    def removeNthFromEnd(self, head: Optional[ListNode], n: int) -> Optional[ListNode]:
        sz = self.len(head)
        n = sz - n  # 0-indexed

        return self.remove(head, n)
        